File: 2025/Day07/Solver.java
package Day07;
import helper.FileReader;
import java.util.*;

public class Solver{
    public static void main(String[] args) throws Exception {
        FileReader fr = new FileReader(args);
        String[] lines = fr.readLines();

        int p1 = 0;
        long p2 = 0;

        int C = lines[0].length();

        long[] beams = new long[C];
        beams[lines[0].indexOf('S')] = 1;

        for (int layer=1; layer<lines.length; layer++){
            long[] temp = new long[C];
            for (int i=0; i<C; i++){
                if (beams[i]==0)continue;
                if (lines[layer].charAt(i)=='.')temp[i] += beams[i];
                else{
                    p1++;
                    if(i+1<C) temp[i+1] += beams[i];
                    if(i-1>=0) temp[i-1] += beams[i];
                }
            }
            beams = temp;
        }

        p2 = Arrays.stream(beams).sum();
        System.out.println("Part 1: "+ p1);
        System.out.println("Part 2: "+ p2);
    }
}

File: 2025/Day07/Solver.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../helper/FileReader.h"

typedef unsigned long long ul;

int main(int argc, char *argv[]) {
    const char *inputFile = (argc > 1) ? argv[1] : "Day6/in.txt";
    
    FileContent lines = readLines(inputFile);

    int p1 = 0;
    ul p2 = 0;

    int C = strlen(lines.content[0]);

    ul* beams = (ul*)malloc(C*sizeof(ul));
    for (int i=0; i<C; i++) beams[i] = lines.content[0][i]!='S'?0:1;

    for (int layer=1; layer<lines.count; layer++){
        ul* temp = (ul*)malloc(C*sizeof(ul));
        for (int i=0; i<C; i++) temp[i] = 0;
        for (int i=0; i<C; i++){
            if (beams[i]==0)continue;
            if (lines.content[layer][i]=='.')temp[i] += beams[i];
            else{
                p1++;
                if(i+1<C) temp[i+1] += beams[i];
                if(i-1>=0) temp[i-1] += beams[i];
            }
        }
        free(beams);
        beams = temp;
    }

    for(int i=0; i<C; i++) p2 += beams[i];
    free(beams);
    freeFileLines(&lines);

    printf("Part 1: %d\n", p1);
    printf("Part 2: %lld\n", p2);
    
    return 0;
}